k8s/src/main/scala/io/buoyant/k8s/IngressCache.scala

```diff
@@ -0,0 +1,128 @@
+package io.buoyant.k8s
+
+import com.twitter.finagle.http.{Request, Response}
+import com.twitter.finagle.{Path, Service}
+import com.twitter.util._
+import java.util.concurrent.atomic.AtomicReference
+
+case class IngressSpec(
+  name: Option[String],
+  namespace: Option[String],
+  fallbackBackend: Option[IngressPath] = None,
+  rules: Seq[IngressPath]
+)
+
+case class IngressPath(
+  host: Option[String] = None,
+  uri: Option[String] = None,
+  namespace: String,
+  svc: String,
+  port: String
+)
+
+object IngressCache {
+  type IngressState = Activity.State[Seq[IngressSpec]]
+
+  private[k8s] def isPrefix(pfx: String, uri: String): Boolean = {
+    val pfxSegments = pfx.split("/")
+    val uriSegments = uri.split("/")
+    pfxSegments.corresponds(uriSegments.take(pfxSegments.size)) { _ == _ }
+  }
+
+  private[k8s] def getMatchingPath(hostHeader: Option[String], requestPath: String, ns: Option[String], ingresses: Seq[IngressSpec]): Option[IngressPath] = {
+    ingresses.flatMap { ingressResource =>
+      val matchingPath = ingressResource.rules.find { rule =>
+        (rule.host, rule.uri, rule.namespace) match {
+          case (Some(host), Some(path), ns) =>
+            isPrefix(path, requestPath) && hostHeader.contains(host)
+          case (Some(host), None, ns) => hostHeader.contains(host)
+          case (None, Some(path), ns) => isPrefix(path, requestPath)
```

Comment:
FWIW, [this](https://github.com/kubernetes/kubernetes/blob/master/pkg/apis/extensions/v1beta1/types.go#L690) says that the path is actually a regex.  We don't necessary have to support that, but it's worth being aware of.